web3/decoding.nim

```diff
@@ -0,0 +1,416 @@
+  let value = decoder.decode(basetype(T))
+  if value in T.low..T.high:
+    T(value)
+  else:
+    raise newException(SerializationError, "value not in range")
+
+proc decode(decoder: var AbiDecoder, T: type enum): T {.raises: [SerializationError].}=
+  let value = decoder.decode(uint64)
+  if value in T.low.uint64..T.high.uint64:
+    T(value)
+  else:
+    raise newException(SerializationError, "invalid enum value")
+
+proc decode(decoder: var AbiDecoder, T: type Address): T {.raises: [SerializationError].} =
+  var bytes: array[sizeof(T), byte]
+  let padding = abiSlotSize - sizeof(T)
+
+  try:
+    bytes[0..<sizeof(T)] =(decoder.read(sizeof(T)))[padding ..< abiSlotSize]
+  except IOError as e:
+    raise newException(SerializationError, "Failed to read address: " & e.msg)
+
+  T(bytes)
+
+proc decode[I](decoder: var AbiDecoder, T: type array[I, byte]): T {.raises: [SerializationError].} =
+  var arr: T
+  let bytes = (arr.len + 31) div 32 * 32
+  let buf = decoder.read(bytes)
+  arr[0..<arr.len] = buf[0..<arr.len]
+
+  checkRightPadding(buf, arr.len, bytes)
+
+  return arr
+
+proc decode(decoder: var AbiDecoder, T: type seq[byte]): T {.raises: [SerializationError].} =
+  let len = decoder.decode(uint64)
+  let bytes = ((len + 31) div 32 * 32).int
+  let buf = decoder.read(bytes)
+
+  checkRightPadding(buf, len.int, bytes)
+
+  return buf[0 ..< len]
+
+proc decode(decoder: var AbiDecoder, T: type string): T {.raises: [SerializationError].} =
+  string.fromBytes(decoder.decode(seq[byte]))
+
+proc decode[T: distinct](decoder: var AbiDecoder, _: type T): T {.raises: [SerializationError].} =
+  T(decoder.decode(distinctBase T))
+
+## When T is dynamic, ABI layout looks like:
+## +----------------------------+
+## | size of the dynamic array |  <-- 32 (optional ONLY for dynamic arrays)
+## +----------------------------+
+## | offset to element 0       |  <-- 32
+## +----------------------------+
+## | offset to element 1       |  <-- 32 + size of encoded element 0
+## +----------------------------+
+## | ...                        |
+## +----------------------------+
+## | encoded element 0         |  <-- item at offset 0
+## +----------------------------+
+## | encoded element 1         |  <-- item at offset 1
+## +----------------------------+
+## | ...                        |
+## +----------------------------+
+##
+## When T is static, ABI layout looks like:
+## +----------------------------+
+## | size of the dynamic array |  <-- 32 (optional ONLY for dynamic arrays)
+## +----------------------------+
+## | element 0                 |  <-- 32
+## +----------------------------+
+## | element 1                 |  <-- 32
+## +----------------------------+
+## | ...                        |
+## +----------------------------+
+## | element N-1               |
+## +----------------------------+
+##
+## The size of the static array is passed as an argument to the decoder.
```

Comment:
🙏 Thank you for doing this, it helps the reader a lot!

Perhaps we should put all the comments inside of the proc, as this is the nim way. It helps with docgen and helps IDEs display in hover text.